### ISSUE TITLE
Use sklearn.ParameterGrid in GKEParallel._partition_param_grid

### ISSUE THREAD

**Issue by pmlandwehr** [GoogleCloudPlatform/ml-on-gcp#21]
`sklearn` can use `ParameterGrid` objects to create all possible combinations of parameters. Notably,  it tolerates lists of dictionaries of paramters (e.g. `[{'a': [1,2,3], 'b': [0, 1]}, {'a':[4,5,6], 'b':[2,3,4]}]`)
This kind of subsetting isn't supported in `GKEParallel`, which stops partitioning jobs across workers as soon as it sees that `param_grid` is a list. As such, we might want to switch to `ParameterGrid` to help with this.

**Comment by dizcology** [comment]
@pmlandwehr: Thanks.  I also felt the current way of distributing the `param_grid` across workers is not ideal.  Are you suggesting that we should expand the grid completely into a list of all combinations, and then distribute that list across the workers?

**Comment by pmlandwehr** [comment]
That was my initial theory, but I'm concerned that this blows up the amount of data you end up pushing to storage. A very compact `param_grid` can contain a lot of combinations. An alternative would be to pass each worker an index and offset. (If you were using compute for this, those arguments could be stored in the metadata.) Each worker could then expand the grid using `ParameterGrid`, iterate to the correct offset, and then only execute on that chunk. On the downside, that's much dirtier than just choosing a particular family of arguments.

One benefit to expanding locally, even if that does blow up the amount of data being moved to and from the cloud, is that you could shuffle all of the hyperparameter configurations in advance. That should help mitigate the impact of one hyperparameter being particularly bad for training on and hopefully guarantee that workers will finish in comparable amounts of time.

**Comment by dizcology** [comment]
Good points on shuffling and the amount of data from expanding the grid.

One possible approach that seems half-way between the two extremes (expanding fully and expanding just one hyperparameter) might be to somehow randomly partition the original `param_grid` into a list of smaller `param_grid`s, and then randomly assign them to the workers.

Then it seems that one should just use `RandomizedSearchCV`.
